### PR TITLE
Implement validations for SignedExtension and UnsignedValidator

### LINKED ISSUE #52: Concrete implementations of UnsignedValidator

### PR DESCRIPTION
**Pull request by radkomih**
**Detailed description**:

### Runtime related changes
* [x] move the validations higher in the hierarchy, so low-level primitives does not depend on higher level modules and validations can be overridden
* [x] implement `non zero signer` validation 
* [x] implement `nonce` validation 
* [x] implement `mortality` validation 
* [x] partial `weight` validation (checks extrinsic length) 
* [x] provide `encode/decode` implementation for some types
* [x] cleanup some `enum` type representations
* [x] change to use `assert` instead of `require` and refactor some test helpers
* [x] cleanup

### SCALE codec related changes
* [x] refactor and use `DecodeOptionWith()`
* [x] refactor the usage of `ToCompact()` 

**Which issue(s) this PR fixes**:
Fixes #52 

**Special notes for your reviewer**:
The rest of the weight validation can be implemented here: https://github.com/LimeChain/gosemble/issues/47

**Checklist**
- [ ] Documentation added
- [x] Tests updated

### CODE REVIEW

**Comment by radkomih** [comment]
need to merge the SCALE changes first

File: runtime/block_builder_apply_extrinsic_test.go [hunk 1]
```diff
@@ -164,44 +128,21 @@ func Test_ApplyExtrinsic_DispatchError_BadProofError(t *testing.T) {
...
-	invalidSignature := types.NewMultiSignature(
-		types.NewEd25519(
-			0xb5, 0x1b, 0xc6, 0xf6, 0xe9, 0xf7, 0x07, 0x90, 0x03, 0xad, 0xc9, 0x8e, 0xd2, 0x16, 0x22, 0xd0, 0xd8, 0x87, 0xd6, 0x8e, 0x86, 0x45, 0xb5, 0xcb, 0x99, 0xe3, 0xd6, 0x08, 0x0a, 0xa8, 0xc9, 0xdd, 0x40, 0x8c, 0xa3, 0x9c, 0x91, 0xd9, 0xc7, 0x0a, 0x49, 0xa3, 0x77, 0xbc, 0x2b, 0x55, 0x04, 0xe3, 0x64, 0x27, 0xe1, 0x84, 0x5b, 0x38, 0x20, 0xc5, 0x8c, 0x95, 0xf1, 0x46, 0xf0, 0xce, 0xc2, 0x03,
-		),
-	)
-
+	extra := newTestExtra(types.NewImmortalEra(), 1, 0) // instead of 0 to make the signature invalid
+	call := newTestCall(0, 0, 0xab, 0xcd)
+	signer := newTestSigner()
+	invalidSignature := newTestSignature(0xb5, 0x1b, 0xc6, 0xf6, 0xe9, 0xf7, 0x07, 0x90, 0x03, 0xad, 0xc9, 0x8e, 0xd2, 0x16, 0x22, 0xd0, 0xd8, 0x87, 0xd6, 0x8e, 0x86, 0x45, 0xb5, 0xcb, 0x99, 0xe3, 0xd6, 0x08, 0x0a, 0xa8, 0xc9, 0xdd, 0x40, 0x8c, 0xa3, 0x9c, 0x91, 0xd9, 0xc7, 0x0a, 0x49, 0xa3, 0x77, 0xbc, 0x2b, 0x55, 0x04, 0xe3, 0x64, 0x27, 0xe1, 0x84, 0x5b, 0x38, 0x20, 0xc5, 0x8c, 0x95, 0xf1, 0x46, 0xf0, 0xce, 0xc2, 0x03)
```
Comment (failfmi) [inline runtime/block_builder_apply_extrinsic_test.go]: Can we switch all signatures written with `0xb5, 0xb6...` to hex strings as this is unreadable?

File: frame/system/extentions/extentions.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
+package system
```
Comment (failfmi) [inline frame/system/extentions/extentions.go]: Shouldn't the folder be named `extensions`?

File: frame/system/extentions/check_non_zero_sender.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+	// TODO:
+	// Not sure when this is possible.
+	// Checks signed transactions but will fail
+	// before this check if the address is all zeros.
+	for _, v := range who.Bytes() {
+		if v != 0 {
+			ok = types.DefaultValidTransaction()
+			return ok, err
+		}
+	}
```
Comment (failfmi) [inline frame/system/extentions/check_non_zero_sender.go]: nit: clear constant zeroAddress and deepEqual it

File: frame/system/extentions/extentions.go (hunk 2 above)
Comment (radkomih) [reply frame/system/extentions/extentions.go]: maybe, I'll change it

File: frame/system/extentions/check_nonce.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+import (
+	"math"
+
+	sc "github.com/LimeChain/goscale"
+	"github.com/LimeChain/gosemble/frame/system"
+	"github.com/LimeChain/gosemble/primitives/types"
+)
+
+type CheckNonce sc.U64
```
Comment (failfmi) [inline frame/system/extentions/check_nonce.go]: nit: this is u32

File: primitives/types/multi_address.go [hunk 5]
```diff
@@ -44,7 +44,7 @@ func (a MultiAddress) AsAccountId() AccountId {
 }
 
 // It's an account index.
-type AccountIndex = sc.U32
+type AccountIndex = sc.U32 // TODO: encode as Compact
```
Comment (failfmi) [inline primitives/types/multi_address.go]: This is not a correct comment, it depends on the specific storage.

**Review by failfmi — DISMISSED** [review]

**Review by failfmi — APPROVED** [review]

**Pull request merged**